agent/agent.go

```diff
@@ -142,107 +142,230 @@ func (a *Agent) Run(ctx context.Context) error {
+		unsent += output.BufferLength()
+	}
+
+	return fmt.Errorf("output plugins unable to send %d metrics", unsent)
+}
+
+func (a *Agent) test(ctx context.Context, wait time.Duration, outputF func(<-chan telegraf.Metric)) error {
 	log.Printf("D! [agent] Initializing plugins")
 	err := a.initPlugins()
 	if err != nil {
 		return err
 	}
 
-	if hasServiceInputs {
-		log.Printf("D! [agent] Starting service inputs")
-		err := a.startServiceInputs(ctx, metricC)
-		if err != nil {
-			return err
-		}
+	log.Printf("D! [agent] Connecting outputs")
+	err = a.connectOutputs(ctx)
+	if err != nil {
+		return err
 	}
 
-	hasErrors := false
-	for _, input := range a.Config.Inputs {
-		select {
-		case <-ctx.Done():
-			return nil
-		default:
-			break
+	inputC := make(chan telegraf.Metric, 100)
+	procC := make(chan telegraf.Metric, 100)
+	outputC := make(chan telegraf.Metric, 100)
+
+	startTime := time.Now()
+
+	log.Printf("D! [agent] Starting service inputs")
+	err = a.startServiceInputs(ctx, inputC)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+
+	src := inputC
+	dst := inputC
+
+	nul := make(chan telegraf.Metric)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for range nul {
 		}
+	}()
 
-		acc := NewAccumulator(input, metricC)
-		acc.SetPrecision(a.Precision())
+	wg.Add(1)
+	go func(dst chan telegraf.Metric) {
+		defer wg.Done()
 
-		// Special instructions for some inputs. cpu, for example, needs to be
-		// run twice in order to return cpu usage percentages.
-		switch input.Config.Name {
-		case "cpu", "mongodb", "procstat":
-			nulAcc := NewAccumulator(input, nulC)
-			nulAcc.SetPrecision(a.Precision())
-			if err := input.Input.Gather(nulAcc); err != nil {
-				acc.AddError(err)
-				hasErrors = true
-			}
+		var wg sync.WaitGroup
+		for _, input := range a.Config.Inputs {
+			wg.Add(1)
+			go func(input *models.RunningInput) {
+				defer wg.Done()
+				jitter := 0 * time.Second
+				interval := a.Config.Agent.Interval.Duration
+
+				// Overwrite agent interval if this plugin has its own.
+				if input.Config.Interval != 0 {
+					interval = input.Config.Interval
+				}
+
+				ticker := NewUnalignedTicker(interval, jitter)
+				defer ticker.Stop()
+				<-ticker.Elapsed()
+
+				switch input.Config.Name {
+				case "cpu", "mongodb", "procstat":
+					nulAcc := NewAccumulator(input, nul)
+					nulAcc.SetPrecision(a.Precision())
+					if err := input.Input.Gather(nulAcc); err != nil {
+						nulAcc.AddError(err)
+					}
+
+					time.Sleep(500 * time.Millisecond)
+				}
+
+				acc := NewAccumulator(input, dst)
+				acc.SetPrecision(a.Precision())
+
+				if err := input.Input.Gather(acc); err != nil {
+					acc.AddError(err)
+				}
+			}(input)
 
-			time.Sleep(500 * time.Millisecond)
-			if err := input.Input.Gather(acc); err != nil {
-				acc.AddError(err)
-				hasErrors = true
-			}
-		default:
-			if err := input.Input.Gather(acc); err != nil {
-				acc.AddError(err)
-				hasErrors = true
-			}
 		}
-	}
+		wg.Wait()
```

Comment:
just clarifying that this is a wait on the internal wg, not the external wg.

Reply:
I created a function for running the inputs in test mode, but yeah it was.